core/arch/arm/kernel/thread_spmc.c

```diff
@@ -246,71 +249,88 @@ static void handle_rxtx_map(struct thread_smc_args *args)
 	tx_pa = args->a2;
 	rx_pa = args->a1;
 
-	if (rxtx_size) {
+	if (rxtx->size) {
 		rc = FFA_DENIED;
 		goto out;
 	}
 
-	rc = map_buf(tx_pa, sz, &tx);
-	if (rc)
-		goto out;
-	rc = map_buf(rx_pa, sz, &rx);
-	if (rc) {
-		unmap_buf(tx, sz);
-		goto out;
+	/*
+	 * If the buffer comes from a SP the address is virtual and already
+	 * mapped.
+	 */
+	if (is_nw_buf(rxtx)) {
+		rc = map_buf(tx_pa, sz, &tx);
+		if (rc)
+			goto out;
+		rc = map_buf(rx_pa, sz, &rx);
+		if (rc) {
+			unmap_buf(tx, sz);
+			goto out;
+		}
+		rxtx->tx = tx;
+		rxtx->rx = rx;
+	} else {
+		if (!virt_to_phys((void *)tx_pa) ||
```

Comment:
i would also assert pa-s are page aligned. Caller should expose plain pages.

Reply:
Done